connectivity/check/check.go

```diff
@@ -168,6 +168,16 @@ type k8sConnectivityImplementation interface {
 	ExecInPod(ctx context.Context, namespace, pod, container string, command []string) (bytes.Buffer, error)
 	ExecInPodWithStderr(ctx context.Context, namespace, pod, container string, command []string) (bytes.Buffer, bytes.Buffer, error)
 	ClusterName() (name string)
+	ListCiliumNetworkPolicies(ctx context.Context, namespace string, opts metav1.ListOptions) (*ciliumv2.CiliumNetworkPolicyList, error)
```

Comment:
This is typically done to allow unit tests run with a mock k8s apiserver implementation. We also have these kind of k8s facing interfaces in other CIlium repos, maybe @aanm could elaborate a bit more here?

Reply:
I think what @ti-mo is referring is that this "catch-all" interface is growing unnecessarily. We should have a single interface for each usage. Also, there is an interface defined in `github.com/cilium/cilium/pkg/k8s/client/clientset/versioned/typed/cilium.io/v2/ciliumnetworkpolicy.go` named `CiliumNetworkPolicyInterface`, so why are we duplicating the information here?

Reply:
> This is typically done to allow unit tests run with a mock k8s apiserver implementation

Of course, but right now, it isn't (as far as I can see), so why have it in the codebase? And when we do want to inject a mock client, it is required to implement 26 methods to satisfy this mega-interface (I reckon many will just be `ErrNotImplemented` stubs initially), which is not a fun job. This is not a criticism on this patch per sé, but on the approach overall, as many consider large interfaces to be an anti-pattern. I assume this is more common throughout the k8s landscape.

As @aanm mentioned, this could be broken down into smaller interfaces that can be composed. For example, the methods you added in this patch could become:

```go
type k8sCNPClient interface {
  GetCiliumNetworkPolicy(...)
  CreateCiliumNetworkPolicy(...)
  ...
}

type k8sCCNPClient interface {
  ListCiliumClusterwideNetworkPolicies(...)
  GetCiliumClusterwideNetworkPolicy(...)
  ...
}
```

Methods like `updateOrCreateCNP()` only act on one group of k8s client methods at a time (`{Get,Create,Update,Delete}CiliumNetworkPolicy(...)`), so this would work. If tomorrow we want to inject a testing dependency into `updateOrCreateCNP()`, only CRUD on CNP needs to be implemented, nothing more. If a function needs to call methods belonging to more than one interface, they can be composed:

```go
type k8sPolicyClient interface {
  k8sCNPClient
  k8sCCNPClient
}
```

and `K8sConnectivityCheck.clients.src` can be changed to a `k8sPolicyClient` instead.

Perhaps splitting up `k8sConnectivityImplementation` and composing it using smaller interfaces would be a good compromise?

Something like this?

```go
type k8sConnectivityImplementation interface {
  k8sServiceClient
  k8sDaemonSetClient
  k8sDeploymentClient
  k8sNamespaceClient
  k8sPodClient
  k8sCNPClient
  k8sCCNPClient
}
```

Reply:
I'll try splitting this up, at least for the policy APIs :-)